### PR TITLE
feat(stream): temporal join support pk prefix join condition

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

-  Loose temporal join restrictions to allow join condition containing only the prefix of the right table's pk. https://github.com/risingwavelabs/risingwave/issues/8123
- This is a necessary step to support index selection for the temporal join in the future.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 1]
```diff
@@ -983,30 +971,76 @@ impl LogicalJoin {
...
+        let mut dist_key_in_order_key_pos = vec![];
+        for d in dist_key {
+            let pos = order_key
+                .iter()
+                .position(|&x| x == d)
+                .expect("dist_key must in order_key");
+            dist_key_in_order_key_pos.push(pos);
+        }
+        // The at least prefix of order key that contains distribution key.
+        let at_least_prefix_len = dist_key_in_order_key_pos
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:991]:
```suggestion
        // The shortest prefix of order key that contains distribution key.
        let shortest_prefix_len = dist_key_in_order_key_pos
```

File: src/stream/src/executor/temporal_join.rs [hunk 2]
```diff
@@ -271,40 +386,42 @@ impl<S: StateStore, const T: JoinTypePrimitive> TemporalJoinExecutor<S, T> {
                 .set(self.right_table.cache.len() as i64);
             match msg? {
                 InternalMessage::Chunk(chunk) => {
+                    // Compact chunk, otherwise the following keys and chunk rows might fail to zip.
+                    let chunk = chunk.compact();
```
Comment (fuyufjh) [inline src/stream/src/executor/temporal_join.rs:390]: Why? `chunk.rows()` should only returns visible rows

**Review by fuyufjh — COMMENTED** [review]
The rest LGTM

File: src/stream/src/executor/temporal_join.rs (hunk 2 above)
Comment (chenzl25) [reply src/stream/src/executor/temporal_join.rs:390]:
```rust
let keys = K::build(&self.left_join_keys, chunk.data_chunk())?;
for ((op, left_row), key) in chunk.rows().zip_eq_debug(keys.into_iter()) {
```

We need to align keys with rows, but `keys` can't handle visible opes.

Comment (BugenZhao) [reply src/stream/src/executor/temporal_join.rs:390]: I guess we can filter out the empty keys for now (but the assumption is really easy to break, we need further encapsulation for this).

Comment (chenzl25) [reply src/stream/src/executor/temporal_join.rs:390]: Yes, this pattern is quite common(`HashJoin`, `TopN`). We can do a refactor for this pattern later.

**Review by st1page — COMMENTED** [review]
LGTM

**Review by st1page — APPROVED** [review]

**Pull request merged**